### PR TITLE
Tentative curve implementation

### PR DESCRIPTION
**Pull request by PalMartin**
First iteration of the curve api implementation of the TARSIS ETC

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "TARSIS_ETC_lib"
+version = "0.1.0"
+edition = "2021"
```
Comment (BatchDrake) [inline Cargo.toml:4]: Update package edition and name to ZarzaETC

File: src/curve/mod.rs [hunk 2]
```diff
@@ -0,0 +1,499 @@
+pub const INFINITY: f64 = f64::INFINITY; // +Inf_f64
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Perhaps reuse f64's `INFINITY` constant instead of declaring a new one.

File: src/curve/mod.rs [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+use std::io;
+use std::io::BufRead;
+use std::collections::binary_heap::Iter;
+use std::error::Error;
+use csv::ReaderBuilder;
+use csv::WriterBuilder;
+
+pub struct Curve {
+    pub m_oob_right: f64,
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Make all these members private and, in case we need to access them, do it by means of getter methods.

File: src/curve/mod.rs [hunk 4]
```diff
@@ -0,0 +1,499 @@
+pub const INFINITY: f64 = f64::INFINITY; // +Inf_f64
+use std::io;
+use std::io::BufRead;
+use std::collections::binary_heap::Iter;
+use std::error::Error;
+use csv::ReaderBuilder;
+use csv::WriterBuilder;
+
+pub struct Curve {
```
Comment (BatchDrake) [inline src/curve/mod.rs:8]: Implement `Default::default()` for the default constructor.

File: src/curve/mod.rs (hunk 3 above)
Comment (BatchDrake) [reply src/curve/mod.rs]: Naming convention: don't use `m_` to mark private members, the name as-is should be fine.

File: src/curve/mod.rs [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    pub m_oob_right: f64,
+    pub m_oob_left: f64,
+    pub m_units_x: String,
+    pub m_units_y: String,
+    pub m_curve_x: Vec<f64>,
+    pub m_curve_y: Vec<f64>,
+}
+
```
Comment (BatchDrake) [inline src/curve/mod.rs:15]: Name types and variants should be in CamelCase.

File: src/curve/mod.rs [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    fn assign(&mut self, other: &Curve) -> ();
+    fn from_existing(&mut self, other: &Curve, y_units: f64) -> ();
+    fn clear(&mut self) -> ();
+    fn debug(&self) -> ();
+    fn load(&mut self, path: &str) -> Result<(Vec<f64> , Vec<f64>), Box<dyn Error>>;
+    fn save(&self, path: &str) -> Result<(), Box<dyn Error>>;
+}
+
+impl CurveOperations for Curve {
+
```
Comment (BatchDrake) [inline src/curve/mod.rs:74]: Coding style: comments in English

File: src/curve/mod.rs [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+use std::error::Error;
+use csv::ReaderBuilder;
+use csv::WriterBuilder;
+
+pub struct Curve {
+    pub m_oob_right: f64,
+    pub m_oob_left: f64,
+    pub m_units_x: String,
+    pub m_units_y: String,
+    pub m_curve_x: Vec<f64>,
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Perhaps implement X, Y curve points as a `BTreeMap` between `NotNan` (from package `ordered_float`) and `f64`. See https://docs.rs/ordered-float/latest/ordered_float/struct.NotNan.html and https://doc.rust-lang.org/std/collections/struct.BTreeMap.html.

File: Cargo.toml (hunk 1 above)
Comment (BatchDrake) [reply Cargo.toml:4]: Name is okay, update edition to 2025.

File: src/curve/mod.rs [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+    curve: BTreeMap<NotNan<f64>,f64>,
+}
+
+pub enum CurveAxis {
+    XAxis,
+    YAxis,
+}
+
+pub trait CurveOperations { 
+    fn default() -> Self;
```
Comment (BatchDrake) [inline src/curve/mod.rs:43]: `default()` is already declared in Rust's `Default` trait, we should rely on it instead of redeclaring it inside `CurveOperations`

File: src/curve/mod.rs [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+    fn from_existing(&mut self, other: &Curve, y_units: f64) -> ();
+    fn clear(&mut self) -> ();
+    fn debug(&self) -> ();
+    fn load_curve(&mut self, path: &str) -> Result<(), Box<dyn Error>>;
+    fn save_curve(&self, path: &str) -> Result<(), Box<dyn Error>>;
+}
+
+impl CurveOperations for Curve {
+
+    fn default() -> Curve {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Follow-up from previous comment, `fn default()` should appear inside a ```impl Default for Curve``` block, something like:

```rust
impl Default for Curve {
    fn default() -> Self {
        // Code for default constructor
    }
}
```

File: src/curve/mod.rs [hunk 10]
```diff
@@ -0,0 +1,483 @@
...
+            units_y: String::new(),
+            curve: BTreeMap::new(),
+        }
+    }
+
+    fn is_oob(&self, x: NotNan<f64>) -> bool {
+        if self.curve.is_empty() { 
+            return true;
+        }
+        if x < *self.curve.keys().next().unwrap() {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: ~Why are we testing `x` against the second point in the curve?~ Also, why not accessing the first element directly by means of `first_key_value()` inside `BTreeMap`?

File: src/curve/mod.rs [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    fn is_oob(&self, x: NotNan<f64>) -> bool {
+        if self.curve.is_empty() { 
+            return true;
+        }
+        if x < *self.curve.keys().next().unwrap() { 
+            return true;
+        }
+        if *self.curve.keys().next_back().unwrap() < x {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Same with `last_key_value()`

File: src/curve/mod.rs [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    fn set_units(&mut self, axis: CurveAxis, units: String) -> () {
+        match axis {
+            CurveAxis::XAxis => self.units_x = Some(units).unwrap(),
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why are we using `Option<String>` here?

File: src/curve/mod.rs [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+        let mut err = 0.0;
+
+        // Check if out of bounds
+        if self.oob_right + self.oob_left != 0.0 {
+            return f64::INFINITY * (self.oob_left + self.oob_right);
+        } 
+
+        if self.curve.is_empty() {
+            return 0.0;
+        }
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Coding style: keep closing and opening curly braces of `else` in the same line.

File: src/curve/mod.rs [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    fn set_units(&mut self, axis: CurveAxis, units: String) -> () {
+        match axis {
+            CurveAxis::XAxis => self.units_x = Some(units).unwrap(),
+            CurveAxis::YAxis => self.units_y = Some(units).unwrap(),
+        }
+    }
+
+    fn integral(&mut self) -> f64 {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: `integral()` should accept a non-mutable reference

File: src/curve/mod.rs [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+            return f64::INFINITY * (self.oob_left + self.oob_right);
+        } 
+
+        if self.curve.is_empty() {
+            return 0.0;
+        }
+
+        else{
+            // Get first point and remove it for the loop
+            let (x0, y0) = self.curve.pop_first().unwrap();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why do we need to remove the first element form the Curve? `integral` should be a read-only operation on the Curve data, it should not alter its contents.

File: src/curve/mod.rs [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+        let mut err = 0.0;
+
+        // Curve is out of bounds
+        if self.oob_right + self.oob_left != 0.0 {
+            return 0.5 * (self.oob_left + self.oob_right);
+        } 
+
+        if self.curve.is_empty() {
+            return 0.0;
+        }
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Ditto: coding style

File: src/curve/mod.rs [hunk 17]
```diff
@@ -0,0 +1,483 @@
...
+                x_prev = *x;
+                y_prev = *y;
+            }
+            // Put the first point back
+            self.curve.insert(x0, y0);
+            return accum;
+        }
+    }
+
+    fn dist_mean(&mut self) -> f64 {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: `dist_mean()` should not accept a mutable reference, it is a read-only operation.

File: src/curve/mod.rs [hunk 18]
```diff
@@ -0,0 +1,483 @@
...
+            return 0.5 * (self.oob_left + self.oob_right);
+        } 
+
+        if self.curve.is_empty() {
+            return 0.0;
+        }
+
+        else{
+            // Get first point and remove it for the loop
+            let (x0, y0) = self.curve.pop_first().unwrap();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Ditto: same issue as in `integral()`, do not modify the contents of the Curve, even if you end up restoring its original state.

File: src/curve/mod.rs [hunk 19]
```diff
@@ -0,0 +1,483 @@
...
+            let x = *self.curve.keys().next().unwrap();
+            self.curve.clear();
+            self.curve.insert(x, k);
+            self.oob_right = k;
+            return;
+        }
+
+        else {
+            // Get first point and remove it for the loop
+            let (x0, y0) = self.curve.pop_first().unwrap();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Ditto: do not modify the curve just to traverse its points.

File: src/curve/mod.rs [hunk 20]
```diff
@@ -0,0 +1,483 @@
...
+
+            let mut dx: NotNan<f64> = NotNan::new(0.0).unwrap();
+
+            for (x, y) in self.curve.iter() {
+                dx = x - x_prev;
+
+                // Kahan summation
+                let term_real = 0.5 * (y + y_prev) * *dx - err;
+                let tmp = accum + term_real;
+                let term_err = tmp-accum;
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Coding style: leave spaces between operands and operators (i.e. `tmp - accum` instead of `tmp-accum`)

File: src/curve/mod.rs [hunk 21]
```diff
@@ -0,0 +1,483 @@
...
+    
+                err = term_err - term_real;
+                accum = tmp;
+
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Where are the points of the Curve being updated?

File: src/curve/mod.rs [hunk 22]
```diff
@@ -0,0 +1,483 @@
...
+        if self.curve.len() == 1 {
+            let x = *self.curve.keys().next().unwrap();
+            self.curve.clear();
+            self.curve.insert(x, k);
+            self.oob_right = k;
+            return;
+        }
+
+        else {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: We do not need to wrap this code in an `else` block. If `self.curve.len() == 1` we would be leaving this method anyways.

File: src/curve/mod.rs [hunk 23]
```diff
@@ -0,0 +1,483 @@
...
+        }
+    }
+
+    fn flip(&mut self) {
+        // Unit flip
+        std::mem::swap(&mut self.units_x, &mut self.units_y);
+        // x & y coordinates flip
+        let mut flip_crv = BTreeMap::new();
+        for (&x, &y) in &self.curve {
+            flip_crv.insert(x,y);
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Actually flip `x` and `y`

File: src/curve/mod.rs [hunk 24]
```diff
@@ -0,0 +1,483 @@
...
+        }
+        self.curve.clear();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why not just replacing `curve` by `flip_crv`?

File: src/curve/mod.rs [hunk 25]
```diff
@@ -0,0 +1,483 @@
...
+        self.curve.append(&mut flip_crv);
+    }
+
+    fn extend_left(&mut self) -> () {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: This actually applies to all functions that do this: `-> ()` is implied if you don't return a value, so let's remove it from all methods to keep things terse.

File: src/curve/mod.rs [hunk 26]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    fn extend_right(&mut self) -> () {
+        if self.curve.is_empty() {
+            return;
+        }
+        self.oob_right = *self.curve.values().next_back().unwrap(); 
+    }
+
+    fn scale_axis_factor(&mut self, axis: CurveAxis, factor: NotNan<f64>) -> () {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: ~Do we really need to give two different names to `scale_axis_factor` and `scale_axis_curve`?~ Okay, it seems that Rust forbids name overloading for whatever reason. Disregard this comment.

File: src/curve/mod.rs [hunk 27]
```diff
@@ -0,0 +1,483 @@
...
+
+    fn extend_right(&mut self) -> () {
+        if self.curve.is_empty() {
+            return;
+        }
+        self.oob_right = *self.curve.values().next_back().unwrap(); 
+    }
+
+    fn scale_axis_factor(&mut self, axis: CurveAxis, factor: NotNan<f64>) -> () { 
+        let mut new_pairs = BTreeMap::new();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why are we creating a new map here for the Y axis? Why not just altering the elements in-place?

File: src/curve/mod.rs [hunk 28]
```diff
@@ -0,0 +1,483 @@
...
+                }
+            }
+            CurveAxis::YAxis => {
+                for (x, y) in self.curve.iter() {
+                    new_pairs.insert(*x, *y * *factor);
+                }
+            }
+        }
+        self.curve.clear();
+        self.curve.append(&mut new_pairs);
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Can't we just assign the new curve to the existing one?

File: src/curve/mod.rs (hunk 26 above)
Comment (BatchDrake) [reply src/curve/mod.rs]: Note to myself: it seems the way to go here is to define a new generic trait (e.g. `Scalable`):

```rust
trait Scalable<T> {
  fn scaleX(&mut self, byWhat: T);
  fn scaleY(&mut self, byWhat: T);
}
```

And the implement `Scalable<NotNan<f64>>` and `Scalable<&Curve>` or something like that.

File: src/curve/mod.rs [hunk 29]
```diff
@@ -0,0 +1,483 @@
...
+
+        self.oob_left *= *factor;
+        self.oob_right *= *factor;
+    }
+
+    fn scale_axis_curve(&mut self, axis: CurveAxis, other: &mut Curve) -> () {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: `other` must not be mutable. We are only reading data on it.

File: src/curve/mod.rs [hunk 30]
```diff
@@ -0,0 +1,483 @@
...
+                    new_pairs.insert(*x * other.get_point(*x), *y);
+                }
+            }
+            CurveAxis::YAxis => {
+                for (x, y) in self.curve.iter() {
+                    new_pairs.insert(*x, *y * other.get_point(*x));
+                }
+            }
+        }
+        self.curve.clear();
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Ditto

File: src/curve/mod.rs [hunk 31]
```diff
@@ -0,0 +1,483 @@
...
+        self.curve.append(&mut new_pairs);
+
+        self.oob_left = other.oob_left;
+        self.oob_right = other.oob_right;
+    }
+
+    fn invert_axis(&mut self, axis: CurveAxis, factor: NotNan<f64>) -> () {
+        match axis {
+            CurveAxis::XAxis => { 
+                for ref mut x in self.curve.keys() {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: I am not so sure we can do this. This will affect the ordering of the elements in the map and break it. According to Rust people, this is undefined behavior (UB). See https://users.rust-lang.org/t/changing-btreemap-key-while-iterating/108624

File: src/curve/mod.rs [hunk 32]
```diff
@@ -0,0 +1,483 @@
...
+                    *x = &(factor / *x);
+                }
+            }
+            CurveAxis::YAxis => {
+                for ref mut y in self.curve.values_mut() {
+                    **y *= &(*factor / **y);
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why do we need `&` here and double de-referencing?

File: src/curve/mod.rs [hunk 33]
```diff
@@ -0,0 +1,483 @@
...
+    }
+
+    fn get_point(&self, x: NotNan<f64>) -> f64 {
+        if self.curve.is_empty() {
+            panic!("The curve is empty.")
+        }
+        if self.curve.contains_key(&x) {
+            return *self.curve.get(&x).unwrap();
+        }
+        else {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: What happens with the out-of-bounds values? Accessing something that is outside the range of defined points is not an error on its own, in fact we have `oob_left` and `oob_right` to specify what values the function takes there.

File: src/curve/mod.rs [hunk 34]
```diff
@@ -0,0 +1,483 @@
...
+            let (&x0, &y0) = self.curve.range(..x).next_back().unwrap_or_else(|| panic!("No previous point found for interpolation"));
+            let (&x1, &y1) = self.curve.range(x..).next().unwrap_or_else(|| panic!("No next point found for interpolation"));
+            let y = y0 + *((x - x0) * (y1 - y0) / (x1 - x0));
+            return y;
+        }
+    }
+
+    fn get_diff(&self, x: NotNan<f64>) -> f64 {
+        let (&prev,_) = self.curve.range(..x).next_back().unwrap_or_else(|| panic!("No previous point found for the provided value"));
```
Comment (BatchDrake) [inline src/curve/mod.rs]: This is not an error.

File: src/curve/mod.rs [hunk 35]
```diff
@@ -0,0 +1,483 @@
...
+        else {
+            // Two cases
+            if x1 != x {
+                // Middle of segment
+                let diff = (y1 - y0) / *((x1 - x0));
+                return diff;
+            }
+            else {
+                // Edge of a segment
+                let (&x2, &y2) = self.curve.range(next..).nth(1).unwrap_or_else(|| panic!("No next-next point found for the provided value"));
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Not an error either.

File: src/curve/mod.rs [hunk 36]
```diff
@@ -0,0 +1,483 @@
...
+                }
+                else {
+                    let diff = (y2 - y0) / *((x2 - x0));
+                    return diff;
+                }
+            }
+        }
+    }
+
+    fn multiply_by(&mut self, other: &mut Curve) {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: `other` should not be mutable.

File: src/curve/mod.rs [hunk 37]
```diff
@@ -0,0 +1,483 @@
...
+        let mut new_pairs = BTreeMap::new();
+        for (x, _) in self.curve.iter() {
+            new_pairs.insert(*x, self.get_point(*x) * other.get_point(*x));
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Note this new curve only has the X points of the existing curve. The new curve should have the union of both `self` and `other`.

File: src/curve/mod.rs [hunk 38]
```diff
@@ -0,0 +1,483 @@
...
+        }
+        self.curve.clear();
+        self.curve.append(&mut new_pairs);
+
+        self.oob_left *= other.oob_left;
+        self.oob_right *= other.oob_right;
+    }
+
+    fn add_curve(&mut self, other: &mut Curve) {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Ditto. `other` should not be mutable and the new curve must have the union of `self` and `other` keys.

File: src/curve/mod.rs [hunk 39]
```diff
@@ -0,0 +1,483 @@
...
+            new_pairs.insert(*x, self.get_point(*x) + other.get_point(*x));
+        }
+        self.curve.clear();
+        self.curve.append(&mut new_pairs);
+
+        self.oob_left += other.oob_left;
+        self.oob_right += other.oob_right;
+    }
+
+    fn add_value(&mut self, val: f64) -> () {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Reminder that we may want to create a generic trait for this operations.

File: src/curve/mod.rs [hunk 40]
```diff
@@ -0,0 +1,483 @@
...
+        self.oob_right += val;
+    }
+
+    fn assign(&mut self, other: &Curve) -> () { 
+        let Some((&own_first_x, &_)) = self.curve.first_key_value() else { todo!() };
+        let Some((&crv_first_x, &_)) = other.curve.first_key_value() else { todo!() };
+        let Some((&own_last_x, &_)) = self.curve.last_key_value() else { todo!() };
+        let Some((&crv_last_x, &_)) = other.curve.last_key_value() else { todo!() };
+        // Nothing to add
+        if crv_first_x == crv_last_x {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Do not compare floating point numbers with `==`. Also, according to the original C++ code, these tests were used to check whether `self` or `other` were empty.

File: src/curve/mod.rs [hunk 41]
```diff
@@ -0,0 +1,483 @@
...
+            return;
+        }
+        // No curve
+        if own_first_x == own_last_x {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Check if this is empty. If it is the case, copy the contents of `other` inside `self`.

File: src/curve/mod.rs [hunk 42]
```diff
@@ -0,0 +1,483 @@
...
+            return;
+        }
+        // Assign middle part
+        for (x, ref mut y) in self.curve.iter() {
+            if other.is_oob(*x) == false {
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Do not compare Booleans against `false`. Use the operator `!` instead.

File: src/curve/mod.rs [hunk 43]
```diff
@@ -0,0 +1,483 @@
...
+                *y = &other.get_point(*x);
```
Comment (BatchDrake) [inline src/curve/mod.rs]: Why are we using `&` here?

File: src/lib.rs [hunk 44]
```diff
@@ -0,0 +1,101 @@
+pub mod curve;
+
```
Comment (BatchDrake) [inline src/lib.rs]: Unit tests should check not only that the method calls do not break, but also that the return value is the one we expect.

File: Cargo.toml (hunk 1 above)
Comment (PalMartin) [reply Cargo.toml:4]: The last version that I can update rust to is rustc 1.84.1 (e71f9a9a9 2025-01-27), which apparently is too old to use the 2025 edition (Error caused by: "this version of Cargo is older than the `2025` edition, and only supports `2015`, `2018`, `2021`, and `2024` editions.").
If I use the 2024 edition with "cargo-features = ["edition2024"]" I get an error caused by this: "the cargo feature `edition2024` requires a nightly version of Cargo, but this is the `stable` channel", which makes me think that the 2024 version isn't stable yet. Should I try to use it?
